cuqi/randomvariable/_ast.py

```diff
@@ -0,0 +1,227 @@
+"""
+CUQIpy specific implementation of an abstract syntax tree (AST) for random variables.
+
+The main purpose of this AST is to allow for a simple and intuitive way of implementing
+algebraic operations on random variables. The AST is used to record the operations
+and then evaluate them when needed by traversing the tree with the __call__ method.
+
+For example, the following code
+
+    x = RandomVariableNode('x')
+    y = RandomVariableNode('y')
+    z = 2*x + 3*y
+
+will create the following AST:
+
+    AddNode(
+        MultiplyNode(
+            ValueNode(2),
+            RandomVariableNode('x')
+        ),
+        MultiplyNode(
+            ValueNode(3),
+            RandomVariableNode('y')
+        )
+    )
+
+which can be evaluated by calling the __call__ method:
+
+    z(x=1, y=2) # returns 8
+
+
+"""
+convert_to_node = lambda x: x if isinstance(x, Node) else ValueNode(x)
+
+class Node:
+    """ Base class for all nodes in the abstract syntax tree.  """
+    def __call__(self, **kwargs):
+        """ Evaluate node at a given parameter value. This will traverse the tree and evaluate given the recorded operations. """
```

Comment:
```suggestion
        """ Evaluate node at a given parameter value. This will traverse the tree and evaluate it given the recorded operations. """
```